internal/k8s/app_protect_configuration.go

```diff
@@ -0,0 +1,459 @@
+	Obj      *unstructured.Unstructured
+	IsValid  bool
+	ErrorMsg string
+}
+
+// AppProtectUserSigEx represents an App Protect User Defined Signature cluster resource
+type AppProtectUserSigEx struct {
+	Obj      *unstructured.Unstructured
+	Tag      string
+	RevTime  time.Time
+	IsValid  bool
+	ErrorMsg string
+}
+
+func (sig *AppProtectUserSigEx) setInvalid(reason string) {
+	sig.IsValid = false
+	sig.ErrorMsg = reason
+}
+
+func (sig *AppProtectUserSigEx) setValid() {
+	sig.IsValid = true
+	sig.ErrorMsg = ""
+}
+
+type appProtectUserSigSlice []*AppProtectUserSigEx
+
+func (s appProtectUserSigSlice) Len() int {
+	return len(s)
+}
+
+func (s appProtectUserSigSlice) Less(i, j int) bool {
+	if s[i].Obj.GetCreationTimestamp().Time.Equal(s[j].Obj.GetCreationTimestamp().Time) {
+		return s[i].Obj.GetUID() > s[j].Obj.GetUID()
+	}
+	return s[i].Obj.GetCreationTimestamp().Time.Before(s[j].Obj.GetCreationTimestamp().Time)
+}
+
+func (s appProtectUserSigSlice) Swap(i, j int) {
+	s[i], s[j] = s[j], s[i]
+}
+
+func createAppProtectPolicyEx(policyObj *unstructured.Unstructured) (*AppProtectPolicyEx, error) {
+	err := ValidateAppProtectPolicy(policyObj)
+	if err != nil {
+		errMsg := fmt.Sprintf("Error validating policy %s: %v", policyObj.GetName(), err)
+		return &AppProtectPolicyEx{Obj: policyObj, IsValid: false, ErrorMsg: failedValidationErrorMsg}, fmt.Errorf(errMsg)
+	}
+	sigReqs := []SignatureReq{}
+	// Check if policy has signature requirement (revision timestamp) and map them to tags
+	list, found, err := unstructured.NestedSlice(policyObj.Object, "spec", "policy", "signature-requirements")
+	if err != nil {
+		errMsg := fmt.Sprintf("Error retrieving Signature requirements from %s: %v", policyObj.GetName(), err)
+		return &AppProtectPolicyEx{Obj: policyObj, IsValid: false, ErrorMsg: failedValidationErrorMsg}, fmt.Errorf(errMsg)
+	}
+	if found {
+		for _, req := range list {
+			requirement := req.(map[string]interface{})
+			if reqTag, ok := requirement["tag"]; ok {
+				timeReq := RevTimes{}
+				if minRev, ok := requirement["minRevisionDatetime"]; ok {
+					minRevTime, err := time.Parse(timeLayout, minRev.(string))
+					if err != nil {
+						errMsg := fmt.Sprintf("Error Parsing time from minRevisionDatetime in %s: %v", policyObj.GetName(), err)
+						return &AppProtectPolicyEx{Obj: policyObj, IsValid: false, ErrorMsg: invalidTimestampErrorMsg}, fmt.Errorf(errMsg)
+					}
+					timeReq.MinRevTime = &minRevTime
+				}
+				if maxRev, ok := requirement["maxRevisionDatetime"]; ok {
+					maxRevTime, err := time.Parse(timeLayout, maxRev.(string))
+					if err != nil {
+						errMsg := fmt.Sprintf("Error Parsing time from maxRevisionDatetime in %s: %v", policyObj.GetName(), err)
+						return &AppProtectPolicyEx{Obj: policyObj, IsValid: false, ErrorMsg: invalidTimestampErrorMsg}, fmt.Errorf(errMsg)
+					}
+					timeReq.MaxRevTime = &maxRevTime
+				}
+				sigReqs = append(sigReqs, SignatureReq{Tag: reqTag.(string), RevTimes: &timeReq})
+			}
+		}
+	}
```

Comment:
Could something like the following be done to reduce nesting?
```suggestion
	if !found {
		return nil, nil
	}
	for _, req := range list {
		requirement := req.(map[string]interface{})
		if reqTag, ok := requirement["tag"]; ok {
			timeReq := RevTimes{}
			if minRev, ok := requirement["minRevisionDatetime"]; ok {
				minRevTime, err := time.Parse(timeLayout, minRev.(string))
				if err != nil {
					errMsg := fmt.Sprintf("Error Parsing time from minRevisionDatetime in %s: %v", policyObj.GetName(), err)
					return &AppProtectPolicyEx{Obj: policyObj, IsValid: false, ErrorMsg: invalidTimestampErrorMsg}, fmt.Errorf(errMsg)
				}
				timeReq.MinRevTime = &minRevTime
			}
			if maxRev, ok := requirement["maxRevisionDatetime"]; ok {
				maxRevTime, err := time.Parse(timeLayout, maxRev.(string))
				if err != nil {
					errMsg := fmt.Sprintf("Error Parsing time from maxRevisionDatetime in %s: %v", policyObj.GetName(), err)
					return &AppProtectPolicyEx{Obj: policyObj, IsValid: false, ErrorMsg: invalidTimestampErrorMsg}, fmt.Errorf(errMsg)
				}
				timeReq.MaxRevTime = &maxRevTime
			}
			sigReqs = append(sigReqs, SignatureReq{Tag: reqTag.(string), RevTimes: &timeReq})
		}
	}
```

Reply:
Actually, `!found` should return the `AppProtectPolicyEx` just without sigReqs, right?

Reply:
yep